v2/pkg/engine/datasource/pubsub_datasource/pubsub_datasource.go

```diff
@@ -321,38 +326,113 @@ func (s *RequestDataSource) Load(ctx context.Context, input []byte, w io.Writer)
+	return node.Ref, inputValueTypeRef
+}
+
+func (p *Planner[T]) extractEventSubject(fieldRef int, subject string) (string, error) {
+	allMatches := argumentTemplateRegex.FindAllStringSubmatch(subject, -1)
+	// if no argument templates are defined, there are only static values
+	if len(allMatches) < 1 {
+		if natsserver.IsValidSubject(subject) {
+			return subject, nil
+		}
+		return "", fmt.Errorf(`subject "%s" is not a valid NATS subject`, subject)
 	}
-	variableTypeRef := p.visitor.Operation.VariableDefinitions[variableDefinition].Type
-	renderer, err := resolve.NewPlainVariableRendererWithValidationFromTypeRef(p.visitor.Operation, p.visitor.Definition, variableTypeRef, string(variableName))
-	if err != nil {
-		return "", err
+	substitutedSubject := subject
+	for templateNumber, matches := range allMatches {
+		if len(matches) != 2 {
+			return "", fmt.Errorf("expected a single matching group for argument template")
+		}
+		// the path begins with ".", so ignore the first empty string element
+		argumentPath := strings.Split(matches[1][1:], ".")
+		fieldNameBytes := p.visitor.Operation.FieldNameBytes(fieldRef)
+		fieldDefinitionRef, ok := p.visitor.Definition.ObjectTypeDefinitionFieldWithName(p.visitor.Walker.EnclosingTypeDefinition.Ref, fieldNameBytes)
+		if !ok {
+			return "", fmt.Errorf(`expected field definition to exist for field "%s"`, fieldNameBytes)
+		}
+		inputObjectDefinitionRef, lastInputValueTypeRef := p.inputObjectDefinitionRefByFieldDefinitionRef(fieldDefinitionRef, []byte(argumentPath[0]))
+		for _, path := range argumentPath[1:] {
+			inputValueNameBytes := []byte(path)
+			if inputObjectDefinitionRef < 0 {
+				return "", fmt.Errorf(`the event subject "%s" defines nested input field "%s" in argument template #%d, but no parent input object was found`, subject, inputValueNameBytes, templateNumber+1)
+			}
+			inputObjectDefinition := p.visitor.Definition.InputObjectTypeDefinitions[inputObjectDefinitionRef]
+			inputObjectDefinitionRef = -1
+			lastInputValueTypeRef = -1
+			for _, inputValueDefinitionRef := range inputObjectDefinition.InputFieldsDefinition.Refs {
+				if !p.visitor.Definition.InputValueDefinitionNameBytes(inputValueDefinitionRef).Equals(inputValueNameBytes) {
+					continue
+				}
+				inputObjectDefinitionRef, lastInputValueTypeRef = p.inputObjectDefinitionRefByInputValueDefinitionRef(inputValueDefinitionRef)
+				break
+			}
+		}
+		if inputObjectDefinitionRef != -1 {
+			return "", fmt.Errorf(`the event subject "%s" defines the final nested input field "%s" in argument template #%d, but it is not a leaf type`, subject, argumentPath[len(argumentPath)-1], templateNumber+1)
+		}
+		if lastInputValueTypeRef == -1 {
+			return "", fmt.Errorf(`the event subject "%s" defines the final nested input field "%s" in argument template #%d, but it does not exist`, subject, argumentPath[len(argumentPath)-1], templateNumber+1)
+		}
+		argumentName := argumentPath[0]
```

Comment:
Can this panic?

Reply:
```
		if len(matches) != 2 {
			return "", fmt.Errorf("expected a single matching group for argument template")
		}
		// the path begins with ".", so ignore the first empty string element
		argumentPath := strings.Split(matches[1][1:], ".")
```

We only have a match if the regex matches, which must be `[a-zA-Z0-9_]` after a period, which means we should be guaranteed to have at least one item in the argument path. If we don't, the regex didn't match, so we'd never get there.

Either way, this tells me that some regex tests would be useful. I'll add them.